### PR TITLE
Bug 4731 Fix Option Expiration Order Tags and Order Event IsAssignment Flag

### PR DESCRIPTION
**Pull request by mchandschuh**
Standardizes order tags and order events related to option exercise/assignment. All orders/events related to option expiry are now prefix with `Option Expired` and individual events include the following:

OTM: OTM
ITM & long: Automatic Exercise
ITM & short: Automatic Assignment

Since the transaction handler combines the order tag and the order event message, we end up with the following:

Option Expired - OTM
Option Expired - Automatic Exercise
Option Expired - Automatic Assignment

The `OrderEvent.IsAssignment` flag was also updated to only be true in ITM assignment cases.

#### Description

* Add OrderRight.GetExerciseDirection(isShort) extension

> Returns the OrderDirection resulting from exercise/assignment of a particular
option right

* Fix option exercise/assignment order tags and order event messages

> The algorithm manager was doing work to determine whether or not the option ended
in exercise or assignment at expiration. This decision should be left for the exercise
model to decide -- from the algorithm manager's perspective, all that matters is that
the option was expired. The DefaultExerciseModel was updated to properly track whether
the option expired with automatic assignment or exercise, dependending on whether or
not we wrote or bought the option (held liability or right, respectively). Updated unit
tests to check for order event counts and order event messages for option exercise cases.

#### Related Issue

Fixes: #4731

#### Motivation and Context

Fixes inaccurate order tags and order event is assignment property

#### Requires Documentation Change

No.

#### How Has This Been Tested?

Existing unit tests in `SecurityPortfolioManagerTests` expanded to test for the new `OrderEvent.Message` values as well as the `IsAssignment` flag value.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Orders/OptionExercise/DefaultExerciseModel.cs [hunk 1]
```diff
@@ -32,52 +32,56 @@ public class DefaultExerciseModel : IOptionExerciseModel
...
+            if (inTheMoney && option.ExerciseSettlement == SettlementType.PhysicalDelivery)
             {
-                return new[] { optionRemoveEvent, addUnderlyingEvent };
-            }
+                var right = option.Symbol.ID.OptionRight;
 
-            return new[] { optionRemoveEvent };
-        }
+                // TODO : Why doesn't this method take into account the directionality of the quantity like other quantity properties/methods
+                var fillQuantity = option.GetExerciseQuantity(order.Quantity);
```
Comment (Martin-Molinero) [inline Common/Orders/OptionExercise/DefaultExerciseModel.cs:68]: > // TODO : Why doesn't this method take into account the directionality of the quantity like other quantity properties/methods

You are referring to the required `right == OptionRight.Call ? fillQuantity : -fillQuantity`? All consumers of `GetExerciseQuantity` seem to be doing the same operation except for `OptionExerciseOrder` that seems like a bug? 🤔 Agree it kinda feels it should be internal to the `GetExerciseQuantity`

**Review by Martin-Molinero — APPROVED** [review]
Nice 👍, think we are missing to update a couple of unit tests.

File: Common/Orders/OptionExercise/DefaultExerciseModel.cs (hunk 1 above)
Comment (mchandschuh) [reply Common/Orders/OptionExercise/DefaultExerciseModel.cs:68]: my thoughts exactly -- throughout lean we use `Quantity` to mean the directional quantity and `AbsoluteQuantity` to mean the magnitude of the quantity. I'll put an additional commit in here refactoring this method to follow this convention.

**Comment by mchandschuh** [comment]
I pushed the changes that fix this immediate issue. I'm continuing to investigating inconsistencies related to `OptionExerciseOrder.Quantity` -- The `TODO` in this commit is among the issues being investigated. I've also decided to _not_ include the `Option.GetExerciseQuantity` method in this commit since it's implementation needed to be logically incorrect in order to maintain existing behavior. I'll be opening a separate PR that includes this new method along with other changes aimed at improving the consistency and usage of `OptionExerciseOrder.Quantity` and explicitly removing the _positive only_ soft requirement on that value (it wasn't being obeyed in most parts of the code anyway).

**Comment by Martin-Molinero** [comment]
>  I've also decided to not include the Option.GetExerciseQuantity method in this commit since it's implementation needed to be logically incorrect in order to maintain existing behavior. I'll be opening a separate PR that includes this new method along with other changes

Sounds good 👍 (issue at https://github.com/QuantConnect/Lean/issues/4766). there are a couple of python unit test which need an update for this PR `PythonOptionTests.FilterReturnsUniverseRegression` & `PythonOptionTests.FilterReturnsListRegression`

**Pull request merged**